File: Daily Exercises/Mar 2021/2021-03-27 : 61. Rotate List/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not k or not head or not head.next: return head
        fast=head
        slow=head
        for i in range(k):
            if fast.next:
                fast=fast.next
            else:
                fast=head
                break
        if fast==head:    
            for j in range(k%(i+1)):
                fast=fast.next
        if fast==head:
            return head
        while fast.next:
            fast=fast.next
            slow=slow.next
        ans=slow.next
        fast.next=head
        slow.next=None
        return ans

File: Daily Exercises/Mar 2021/2021-03-27 : 61. Rotate List/61. Rotate List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(k==0 || head==null || head.next==null){
            return head;
        }
        ListNode fast=head,slow=head;
        int i=0;
        for(;i<k;i++){
            if(fast.next!=null){
                fast=fast.next;
            }else{
                fast=head;
                break;
            }
        }
        if(fast==head){
            for(int j=0;j<k%(i+1);j++){
            fast=fast.next;
            }
        }
        if(fast==head){
            return head;
        }
        while(fast.next!=null){
            fast=fast.next;
            slow=slow.next;
        }
        ListNode ans=slow.next;
        fast.next=head;
        slow.next=null;
        return ans;
    }
}
